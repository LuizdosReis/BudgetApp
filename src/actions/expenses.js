import uuid from 'uuid';

export const add = ({
  description = '', category = null, createAt = 0, amount = 0,
} = {}) => ({
  type: 'ADD_EXPENSE',
  expense: {
    id: uuid(),
    description,
    category,
    createAt,
    amount,
  },
});

export const remove = ({ id } = {}) => ({
  type: 'REMOVE_EXPENSE',
  id,
});

export const edit = (id, updates) => ({
  type: 'EDIT_EXPENSE',
  id,
  updates,
});
