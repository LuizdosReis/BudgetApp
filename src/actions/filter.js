export const setTextFilter = ({text = ''} = {}) => ({
    type: 'SET_TEXT_FILTER',
    text
})

export const sortByDescription = () => ({
    type: 'SORT_BY_DESCRIPTION'
})

export const sortByType = () => ({
    type: 'SORT_BY_TYPE',
})