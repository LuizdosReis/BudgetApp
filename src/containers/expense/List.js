import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import getVisibleExpenses from '../../selectors/expenses';
import Header from '../../pages/Header';
import Item from './Item';
import FilterExpenses from './Filter';

export const List = ({ expenses }) => (
  <div id="layout">
    <div>
      <Header title="Despesas" />
      <main className="content">
        <div className="main__header">
          <Link to="/expenses/new">
            <button type="submit" className="pure-button pure-button-primary">
              Add Despesas
            </button>
          </Link>
          <FilterExpenses />
        </div>
        {expenses.length === 0 ? (
          <p>Não há despesas</p>
        ) : (
          <div>
            <table className="pure-table table">
              <thead>
                <tr>
                  <th>Descrição</th>
                  <th>Valor</th>
                  <th>Categoria</th>
                  <th>Data</th>
                </tr>
              </thead>
              <tbody>
                {expenses.map(expense => (
                  <Item key={expense.id} {...expense} />
                ))}
              </tbody>
            </table>
          </div>
        )}
      </main>
    </div>
  </div>
);

const mapStateToProps = state => ({
  expenses: getVisibleExpenses(state.expenses, state.expensesFilter),
});

export default connect(
  mapStateToProps,
  undefined,
)(List);
