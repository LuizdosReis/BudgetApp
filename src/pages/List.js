import React from 'react';
import PropTypes from 'prop-types';
import Category from './Category';

const List = ({ categories, handleOpenModal }) => (
  <div>
    <table className="pure-table table">
      <thead>
        <tr>
          <th>Descrição</th>
          <th>Tipo</th>
          <th>Remover</th>
        </tr>
      </thead>
      <tbody>
        {categories.map(category => (
          <Category key={category.id} {...category} handleOpenModal={handleOpenModal} />
        ))}
      </tbody>
    </table>
  </div>
);

List.propTypes = {
  categories: PropTypes.arrayOf(
    PropTypes.shape({
      description: PropTypes.string.isRequired,
      type: PropTypes.string.isRequired,
      id: PropTypes.number.isRequired,
    }),
  ).isRequired,
  handleOpenModal: PropTypes.func.isRequired,
};

export default List;
