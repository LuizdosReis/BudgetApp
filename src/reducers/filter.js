const filterReducerDefaultState = {
    text: '',
    sortBy: 'id'
}

export default (state = filterReducerDefaultState, action) => {
    switch(action.type){
        case 'SET_TEXT_FILTER':
            return {
                ...state,
                text: action.text
            }
        case 'SORT_BY_DESCRIPTION':
            return {
                ...state,
                sortBy: 'description'
            }
        case 'SORT_BY_TYPE':
            return {
                ...state,
                sortBy: 'type'
            }
        default:
            return state;
    }
}
