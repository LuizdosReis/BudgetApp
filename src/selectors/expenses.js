import moment from 'moment';

export default (expenses, {
  startDate, endDate, description, sortBy,
}) => expenses
  .filter((expense) => {
    const createAtMoment = moment(expense.createAt);
    const startDateMatch = startDate ? createAtMoment.isSameOrAfter(startDate, 'day') : true;
    const endDateMatch = endDate ? createAtMoment.isSameOrBefore(endDate, 'day') : true;
    const descriptionMatch = description
      ? expense.description.toLowerCase().includes(description.toLowerCase())
      : true;
    return startDateMatch && endDateMatch && descriptionMatch;
  })
  .sort((a, b) => {
    if (sortBy === 'description') {
      return a.description < b.description ? -1 : 1;
    }
    if (sortBy === 'amount') {
      return a.amount < b.amount ? -1 : 1;
    }
    return a.createAt < b.createAt ? -1 : 1;
  });
