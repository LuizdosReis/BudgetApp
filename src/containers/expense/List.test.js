import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { List } from './List';
import expenses from '../../fixtures/expenses';
import { filters } from '../../fixtures/filter';

test('should render List with expenses', () => {
  const wrapper = shallow(<List expenses={expenses} filter={filters} />);
  expect(toJson(wrapper)).toMatchSnapshot();
});

test('should render List with empty message', () => {
  const wrapper = shallow(<List expenses={[]} filter={filters} />);
  expect(toJson(wrapper)).toMatchSnapshot();
});
