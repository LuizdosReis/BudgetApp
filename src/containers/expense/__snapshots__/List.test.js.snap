// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render List with empty message 1`] = `
<div
  id="layout"
>
  <div>
    <Header
      title="Despesas"
    />
    <main
      className="content"
    >
      <div
        className="main__header"
      >
        <Link
          replace={false}
          to="/expenses/new"
        >
          <button
            className="pure-button pure-button-primary"
            type="submit"
          >
            Add Despesas
          </button>
        </Link>
        <Connect(Filter) />
      </div>
      <p>
        Não há despesas
      </p>
    </main>
  </div>
</div>
`;

exports[`should render List with expenses 1`] = `
<div
  id="layout"
>
  <div>
    <Header
      title="Despesas"
    />
    <main
      className="content"
    >
      <div
        className="main__header"
      >
        <Link
          replace={false}
          to="/expenses/new"
        >
          <button
            className="pure-button pure-button-primary"
            type="submit"
          >
            Add Despesas
          </button>
        </Link>
        <Connect(Filter) />
      </div>
      <div>
        <table
          className="pure-table table"
        >
          <thead>
            <tr>
              <th>
                Descrição
              </th>
              <th>
                Valor
              </th>
              <th>
                Categoria
              </th>
              <th>
                Data
              </th>
            </tr>
          </thead>
          <tbody>
            <Item
              amount={32.5}
              category="Life"
              createAt={0}
              description="Gum"
              id={1}
              key="1"
            />
            <Item
              amount={32.5}
              category="Home"
              createAt={-86400000}
              description="Rent"
              id={2}
              key="2"
            />
            <Item
              amount={32.5}
              category="Bank"
              createAt={86400000}
              description="Credit car"
              id={3}
              key="3"
            />
          </tbody>
        </table>
      </div>
    </main>
  </div>
</div>
`;
